waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -203,36 +167,27 @@ proc addPeer*(pm: PeerManager, remotePeerInfo: RemotePeerInfo, proto: string) =
     return
   
   debug "Adding peer to manager", peerId = remotePeerInfo.peerId, addr = remotePeerInfo.addrs[0], proto = proto
-  
+
   # ...known addresses
   for multiaddr in remotePeerInfo.addrs:
-    pm.peerStore.addressBook.add(remotePeerInfo.peerId, multiaddr)
-  
+    # TODO: unsure why this does not work. understand and fix
+    #pm.peerStore[AddressBook][remotePeerInfo.peerId].add(multiaddr)
+    pm.peerStore[AddressBook][remotePeerInfo.peerId] = pm.peerStore[AddressBook][remotePeerInfo.peerId] & multiaddr
+
   # ...public key
   var publicKey: PublicKey
   discard remotePeerInfo.peerId.extractPublicKey(publicKey)
 
-  pm.peerStore.keyBook[remotePeerInfo.peerId] = publicKey
+  pm.peerStore[KeyBook][remotePeerInfo.peerId] = publicKey
 
-  # ...associated protocols
-  pm.peerStore.protoBook.add(remotePeerInfo.peerId, proto)
+  # TODO: Same. Why this does not work?
+  #pm.peerStore[ProtoBook][remotePeerInfo.peerId].add("d")
+  pm.peerStore[ProtoBook][remotePeerInfo.peerId] = pm.peerStore[ProtoBook][remotePeerInfo.peerId] & proto
 
   # Add peer to storage. Entry will subsequently be updated with connectedness information
   if not pm.storage.isNil:
     pm.storage.insertOrReplace(remotePeerInfo.peerId, pm.peerStore.get(remotePeerInfo.peerId), NotConnected)
 
-proc selectPeer*(pm: PeerManager, proto: string): Option[RemotePeerInfo] =
-  # Selects the best peer for a given protocol
-  let peers = pm.peers.filterIt(it.protos.contains(proto))
-
-  if peers.len >= 1:
-     # TODO: proper heuristic here that compares peer scores and selects "best" one. For now the first peer for the given protocol is returned
-    let peerStored = peers[0]
-
-    return some(peerStored.toRemotePeerInfo())
-  else:
-    return none(RemotePeerInfo)
-
 proc reconnectPeers*(pm: PeerManager,
```

Comment:
I'm unsure if we should keep this reconnection logic, at least in the medium term: this was added before discovery methods, mainly to reconnect fleet nodes to each other after a restart. The main difference between the previous peer store and the underlying libp2p one, is that the previous peer store _only_ contained peers that we initiated a connection to, whereas the libp2p peer store also contains all nodes connecting to us. This implies that we only attempted to reconnect to peers that we explicitly connected to before (e.g. using the `connectToNodes` API call). Because GossipSub has a mandatory backoff time before GRAFTing a PRUNEd connection (implicitly then reconnecting after a restart), we had to wait before attempting to reconnect to each node. IIRC this can block for quite a while when the node starts (and maybe much more so now that _all_ nodes will be attempted with reconnection). Now that nodes generally use DNS discovery on boot to reconnect to nodes, perhaps the whole idea of peer persistence and reconnection should be rethought?

Reply:
Good point. If that's fine I will just change what is needed to not break anything and leave this for another PR.

> perhaps the whole idea of peer persistence and reconnection should be rethought?

Yep. Regarding reconnections, I think `reconnectPeers` can be rescoped to be `runPeeringStrategy` (not happy with the name) with the following functionalities:
* Infinite loop.
* If connectedPeers is below a threshold, try to connect to new peers, from the peerstore.
* It respects the backoff (as its done)
* Some criteria TBD can be added on which peers to select, like score, keeping a given inbound/outbound ratio, keeping a store/relay/etc nodes ratio, etc.
* Forces disconnections if needed.
* Its totally agnostic of peer discovery.
* And does not add peers to the peer store as done now: `pm.addPeer(remotePeerInfo, proto)`

Beyond `runPeeringStrategy`, we will have `runPrunningStrategy` that will remove peers from the peerstore and `runDiscoveryStrategy` that will populate the peerstore with new peers. Names TBD.

Reply:
> Good point. If that's fine I will just change what is needed to not break anything and leave this for another PR.

~Absolutely. I think, though, this PR will change the current behaviour, unless I'm missing something. Previously on startup a peer would reconnect _only_ to peers that were statically added or connected via DNS discovery. Now _all_ peers will both be persisted and a reconnection attempted.~

Changed my thinking halfway through typing this: the peers we _persist_ will still only be the ones added via `addPeers()` (i.e. the ones we ourselves statically configured, discovered via DNS, etc) and not include all peers in the peer store (which would include incoming connections too)? My worry was that we'll now keep persisting all these peers and reattempt connecting to all of them forever (since the peer storage is never truncated/pruned as it stands).

Reply:
> we'll now keep persisting all these peers and reattempt connecting to all of them forever

As I can see in [kibana](https://kibana.infra.status.im/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-7d,to:now))&_a=(columns:!(message,severity),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:ffcc22b0-0116-11ed-9719-cdd3b483481c,key:fleet,negate:!f,params:!(status.prod),type:phrases),query:(bool:(minimum_should_match:1,should:!((match_phrase:(fleet:status.prod)))))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:ffcc22b0-0116-11ed-9719-cdd3b483481c,key:severity,negate:!f,params:(query:6),type:phrase),query:(match_phrase:(severity:6))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:ffcc22b0-0116-11ed-9719-cdd3b483481c,key:logsource,negate:!f,params:(query:node-01.do-ams3.status.prod),type:phrase),query:(match_phrase:(logsource:node-01.do-ams3.status.prod)))),index:ffcc22b0-0116-11ed-9719-cdd3b483481c,interval:auto,query:(language:kuery,query:'%22Reconnecting%20peers%22%20OR%20%22Dialing%20peer%22'),sort:!(!('@timestamp',desc)))) we are already attempting to connect to all peers forever, so after checking it I would say the behaviour is the same. But I have this in mind and planning to further investigate it, also related to the issue in the disv5 loop that we discussed. We are constanly discovering peers, connecting to them. But since we don't have that many peers, we are always trying to connect to the same 10-20 nodes.

> which would include incoming connections too

Yes! Will address all this in another PR, with a new field indicating `Direction` in/out. Then that would be used to complete https://github.com/waku-org/nwaku/issues/1206 (showing number of in/out connections)